### DISCUSSION TITLE
The mapping between zeroshot topic and its name (label) changes after running _sort_mappings_by_frequency() in _bertopic.py?

### DISCUSSION THREAD

**Discussion by harshalcoep** [MaartenGr/BERTopic#2103] [Q&A]
I am using bertopic modeling with zeroshot classification step to assign pre-defined topics (from zeroshot_topic_list) to arxiv abstracts. Here is the code snippet.

`from datasets import load_dataset
from bertopic import BERTopic

dataset = load_dataset("CShorten/ML-ArXiv-Papers")["train"]
docs = dataset["abstract"][:5_000]

zeroshot_topic_list = ["Clustering", "Topic Modeling", "Large Language Models"]

topic_model = BERTopic(
    embedding_model="thenlper/gte-small", 
    min_topic_size=15,
    zeroshot_topic_list=zeroshot_topic_list,
    zeroshot_min_similarity=.85,
    verbose=True
)
topics, _ = topic_model.fit_transform(docs)
`

**Please note that I have modifed _bertopic.py with extra logging statements to see how zeroshot classification step assigns pre-defined labels from zeroshot_topic list to the abstracts. The size of the zeroshot_topic_list in our code is 3.** 

As you can see from the logs below, the Zeroshot Step 1 assigns 397 out of 5000 abstracts with a label from out zeroshot_topic_list. This label information is stored in the Topic column of the dataframe. The rest 4603 abstracts are then sent for clustering which assigns a label to the remaining documents. This label information is stored in the Topic column of the dataframe. The topics from clustering and zeroshot steps are combined using _combine_zeroshot_topics() function. In the dataframe, zeroshot topics are inserted between the outlier topic (that may or may not exist) and the rest of the topics from clustering. Therefore, after the combine step (Zeroshot Step 2), the Topic column in the dataframe has values 0, 1 and 2 corresponding to three zeroshot topics from the list. You can verify in the log (shown below) that the total number of documents assigned to these topics is  281+82+34=397. 

Then comes a sorting function _sort_mappings_by_frequency() that sorts the dataframe based on frequency of documents assigned to each topic. I think that this sorting steps destroys the mapping between zeroshot topics and their names (labels). After, the sorting step, the total number of documents assigned to the topics 0, 1 and 2 changed to 281+252+193=726. After training the model, I used get_topic_info() to print the topics and their corresponding labels. I observed that 0, 1 and 2 still refers to labels from zeroshot_topic_list, but the total documents assigned to these topics are 726. The total documents assigned to zeroshot_topic_list in the zeroshot classification step was 397 and not 726. **This could be a bug?** 

`2024-07-30 14:35:41,537 - BERTopic - Embedding - Transforming documents to embeddings.
Batches: 100%|██████████| 157/157 [00:22<00:00,  6.96it/s]
2024-07-30 14:36:05,695 - BERTopic - Embedding - Completed ✓
2024-07-30 14:36:05,696 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm
2024-07-30 14:36:38,935 - BERTopic - Dimensionality - Completed ✓
2024-07-30 14:36:38,936 - BERTopic - Zeroshot Step 1 - Finding documents that could be assigned to either one of the zero-shot topics
2024-07-30 14:36:39,031 - BERTopic - Zeroshot Step 1 -  Identified labels for 397 documents ✓
2024-07-30 14:36:39,032 - BERTopic - Zeroshot Step 1 -  Determine labels for 4603 documents ✓
2024-07-30 14:36:39,034 - BERTopic - Zeroshot Step 1 - Completed ✓
2024-07-30 14:36:56,922 - BERTopic - Cluster - Start clustering the reduced embeddings
2024-07-30 14:36:57,109 - BERTopic - Cluster - Clustered 4603 documents✓
2024-07-30 14:36:57,110 - BERTopic - Cluster - Completed ✓
2024-07-30 14:36:57,111 - BERTopic - Zeroshot Step 2 - Combining topics from zero-shot topic modeling with topics from clustering...
2024-07-30 14:36:57,128 - BERTopic - Zeroshot Step 2 -  Topic
0    281
1     82
2     34
Name: count, dtype: int64✓
2024-07-30 14:36:57,130 - BERTopic - Zeroshot Step 2 -  Topic
0    281
1     82
2     34
Name: count, dtype: int64✓
2024-07-30 14:36:57,131 - BERTopic - Zeroshot Step 2 - Completed ✓
2024-07-30 14:36:57,133 - BERTopic - After combining Topic
0    281
1     82
2     34
Name: count, dtype: int64✓
2024-07-30 14:36:57,138 - BERTopic - After sorting Topic
0    281
1    252
2    193
Name: count, dtype: int64✓
2024-07-30 14:36:57,139 - BERTopic - Representation - Extracting topics from clusters using representation models.
2024-07-30 14:36:57,672 - BERTopic - Representation - Topic
0    281
1    252
2    193
Name: count, dtype: int64
2024-07-30 14:36:57,673 - BERTopic - Representation - Completed ✓

First few rows (and columns) after running topic_model.get_topic_info() 
0  -1  1616  -1_the_of_and_to  
1  0  281  Clustering  
2  1  247  Topic Modeling  
3  2  122  Large Language Models
 `

**Comment by MaartenGr** [comment]
@ianrandman I think this may relate to the previously linked problem wrt zero-shot modeling (#2102)

**Reply by harshalcoep** [reply]
Exactly @MaartenGr and @ianrandman , I encountered a similar issue as mentioned in #2102. To investigate, I added logging statements in _bertopic.py to pinpoint the (possible) bug. I believe the problem lies in _sort_mappings_by_frequency(). As the logs show, everything works as expected until this function is called. I can also send you the modified _bertopic.py with the additional logging statements.
